### PR TITLE
Add default template to the ContentControl, remove theme specific templates

### LINKED ISSUE #4614: Discussion :  Application.Styles required or Window will not Render Content
### Default Styles Required for Window to Render?

This discussion relates to a newbie mistake that could frustrate people who are trying to port a WPF application to Avalonia.

While attempting to port a portion of my application that consists of 7assemblies, I became eager to finally see zero compilation errors so I could run the application.  Since this was a port, I did not start the process using the Avalonia Visual Studio project templates.  However, I did create a side project using the templates and referred to it to test out approaches and to understand what needed to be changed in the WPF app.

Once I executed the application, the only thing I saw was a black Window.  Black not because I set the background color to black, but rather due to an error on my part.  As I walked line by line comparing the port with a running Avalonia application, I just couldn't spot what was missing.

Eventually I realized that the black Window was caused by the fact that I did not include default styles as below.  Since I was porting a decent amount of code for this portion of my application, I had a lot of places where I blamed myself for making mistakes.

```
<Application
    xmlns="https://github.com/avaloniaui"
    x:Class="Oceanside.VirtualKeyboard.Sample.App.App"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
  <Application.Styles>
    <StyleInclude Source="avares://Avalonia.Themes.Default/DefaultTheme.xaml"/>
    <StyleInclude Source="avares://Avalonia.Themes.Default/Accents/BaseLight.xaml"/>
  </Application.Styles>
</Application>
```

I didn't include those lines because I didn't think any of them would be needed.  My application has data templates for all of my view models, the Window itself has no decorations and I thought all style was under my control.

> Technically, when I first ran the application, nothing at all showed.  In an effort to debug the problem, I removed the transparency from the Window and set it to Blue.  The Window would render black, almost like it wasn't rendering anything at all.

### Sample Project
[ContentWillNotShow.zip](https://github.com/AvaloniaUI/Avalonia/files/5171777/ContentWillNotShow.zip)

### Screen Shot of Running Sample App having Styles Commented Out
image

### Expectations
Since default styles are not required in WPF, it took me a while to realize what was causing a blank, control-less, black Window.  Its hardest to spot when you are porting existing WPF code to Avalonia.  I will leave it up to your team to decide if rendering a black Window (or not rendering at all) is by design when default styles are not included.  Maybe mentioning the requirement [here](https://avaloniaui.net/docs/styles/styles) will help and or throwing an exception with a missing styles message?

### PR DESCRIPTION
**Pull request by maxkatz6**
## What does the pull request do?

Created this PR as I wanted to do something different in my free time, but what might improve some existing problems.
CC @grokys as I want to hear your opinion about these changes.

There were couple of different discussions before:
1. There is a [good opinion](https://github.com/AvaloniaUI/Avalonia/issues/4614) that Avalonia app should provide at least the bare minimum templates without any additional themes. We don't want to duplicate themes/templates in the C# code as "yet simpler version of simple theme", but we can at least move some common templates from XAML to Avalonia core package.
2. There is an annoying (at least for me) issue with UserControl that forced as to use [StyleInclude here](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/FluentControls.xaml#L78). It makes resources dictionary merging less efficient. And when add transformer to eliminate StaticResource where it's possible (by referencing resource reference directly if it's in the same dictionary), this Styles+StyleInclude would make it impossible to work efficiently.

Now, adding a default value to the ContentControl.Template solves quite a lot of that:
- ContentControl obviously will have everything it needs to render without a theme.
- UserControl too, but also we can remove this Styles+StyleInclude combination, simplifying our default themes (and third party ones too).
- More complex control like Window, TopLevel, Button... now also have a bare minimum template too (since they inherit ContentControl). Instead of showing transparent window. But still should be restyled to get full functionality.
- Slightly minimal perf improvement as well I suppose, don't even want to measure if.
- Headless tests without any theme attached now make sense. At least minimal test with Window and Border can work now, as Window will have a template to host a content.

With all of this, ControlCatalog (with some unrelated modifications) can run without fluent nor simple theme:
image

image

## Can it be improved futher?

Technically, we can do the same for ItemsControl (which will cover ListBox as well), so it doesn't need to be restyled by every third-party theme.
ScrollViewer with only ScrollViewerPresenter inside, no bars, also can be done in this way, but it 100% will be reimplemented by every third-party theme, so not that many reasons. Possibly only headless.
Other controls are too complicated for that, and also will be restyled in the end anyway.

## Fixed issues

Fixes #4614

### CODE REVIEW

**Comment by tomenscape** [comment]
I like this change. We are re-theming everything in our application and #4614 was a roadblock that we hit very early on. `PopupRoot` has the same problem I believe, which left us scratching our heads as tooltips never appeared, even though we had a theme for `ToolTip`.

We can't even identify all the special templates that we need, and over time new ones could easily be introduced.

The solution we ended up with was including the simple theme before our own theme, so that we can be certain that we have a template for every obscure low-level control. It would be great if we could avoid doing this.

**Comment by maxkatz6** [comment]
@tomenscape popuproot should minimally work, as it's content control as well. 

Saying that, tests didn't like this change at all.

**Comment by grokys** [comment]
I've been thinking about this, here are my thoughts so far:

## Cons

> There is a [good opinion](https://github.com/AvaloniaUI/Avalonia/issues/4614) that Avalonia app should provide at least the bare minimum templates without any additional themes

I'm not sure that's the conclusion I come to when reading that thread: that issue seems more to be asking for _any_ indication of why a control isn't displaying due to a missing template. At the beginning of the project for example, I considered draing a red cross for controls without a template (never did implement this, probably for the best ;) ).

> We don't want to duplicate themes/templates in the C# code as "yet simpler version of simple theme", but we can at least move some common templates from XAML to Avalonia core package.

Yeah, the question would be: how far do we go? Because the logical conclusion of this is that we always ship something like the default theme as default templates.

## Pros

- > we can remove this Styles+StyleInclude combination, simplifying our default themes (and third party ones too).
- > Headless tests without any theme attached now make sense.
- Basically everything you listed!

## Conclusion

I'm still not completely decided on this mainly because of the question "How far do we take this?". _But_, given that `ContentControl` (and derived controls) are such a fundamental building block, and given that 99% of the time the `ContentControl` template will be exactly the same, regardless of the theme being used, I'm tempted to say that this makes sense _in the case, and only in the case of `ContentControl`_.

File: src/Avalonia.Controls/ContentControl.cs [hunk 1]
```diff
@@ -43,6 +43,23 @@ public class ContentControl : TemplatedControl, IContentControl, IContentPresent
         static ContentControl()
         {
             ContentProperty.Changed.AddClassHandler<ContentControl>((x, e) => x.ContentChanged(e));
+            TemplateProperty.OverrideDefaultValue<ContentControl>(new FuncControlTemplate((c, ns) =>
+            {
+                var presenter = new ContentPresenter
+                {
+                    [~BackgroundProperty] = c.GetObservable(BackgroundProperty).ToBinding(),
```
Comment (grokys) [inline src/Avalonia.Controls/ContentControl.cs:50]: I'm not sure if this will be a problem, but I think this means that the subscriptions to the `ContentControl` properties will never be released. So if the default value is used, and then a template is supplied, the original default template subscriptions won't be removed.

Comment (grokys) [reply src/Avalonia.Controls/ContentControl.cs:50]: I think we should probably use `TemplateBinding`s here to fix this.

Comment (grokys) [reply src/Avalonia.Controls/ContentControl.cs:50]: `TemplateBinding` has now been moved to Avalonia.Base by #11483.

**Comment by Gillibald** [comment]
Our control unit tests are full of control template factories that create bare minimum templates that only contain parts that are mandatory for the control being used. So having such factory as part of the actual control would help to speed up testing and makes it easier to tests controls that rely on specific parts in their control template.

TextBox for example will need a ScrollViewer and a TextPresenter but the ScrollViewer itself isn't functional without a template. So I end up writing a template factory for ScrollViewer and for TextBox just to be able to test. All this requires internal knowledge of the controls being used.

File: tests/Avalonia.Controls.UnitTests/ComboBoxTests.cs [hunk 2]
```diff
@@ -430,7 +430,8 @@ public void FlowDirection_Of_RectangleContent_Updated_After_OpenPopup()
                     {
                         new ComboBoxItem()
                         {
-                            Content = parentContent.Child
+                            Content = parentContent.Child,
+                            Template = null // ugly hack, so we can "attach" same child to the two different trees
```
Comment (maxkatz6) [inline tests/Avalonia.Controls.UnitTests/ComboBoxTests.cs:434]: This hack replicates old test behavior, when there was no template on the ComboBoxItem. Now there is one always.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0039008-beta`. (feed url: https://nuget-feed-all.avaloniaui.net/v3/index.json) [PRBUILDID]

**Comment by MrJul** [comment]
I had a look at the code and it's looking good to me.

What's the final consensus about merging this?

**Comment by maxkatz6** [comment]
@MrJul after discussion we agreed on removing ItemsControl template and not adding more than ContentControl template. So, this PR should be mergeable now.

**Review by MrJul — APPROVED** [review]
